### PR TITLE
Define a queue metrics reporter interface 

### PR DESCRIPTION
**Pull request by fearful-symmetry**
This is the result of a brief conversation with @faec about how to add metrics reporting to the queue. After going over some of the requirements in https://github.com/elastic/beats/issues/31113, I just made a first attempt at defining an interface for what metrics would actually be reported. 

I added a simple "reporter" to the `Queue` interface, but I'm not sure we might actually want to do it that way. Whatever calls that `Metrics()` function will have to be handled by metrics reporter that lives in its own thread, so maybe the `Queue` interface should return some kind of handler or something, but I'm not sure.

### CODE REVIEW

File: libbeat/publisher/queue/queue.go
Comment (cmacknz) [inline libbeat/publisher/queue/queue.go:40]:
```suggestion
	QueueFullCount uint64
```

The one below this is `QueueIsFull` so this seems more consistent. Either Limit or Full works as long as it is used consistently.

File: libbeat/publisher/queue/queue.go
Comment (faec) [inline libbeat/publisher/queue/queue.go:37]: Actually I'd rather _these_ metrics be unambiguous and consistently implementable. Generic / high-level / comprehensible ones should ideally happen in the metrics code in a queue-agnostic way (rather than being duplicated in different queues where the semantics can subtly diverge). So my main comment below is about making the metrics more concrete and low-level 😅

File: libbeat/publisher/queue/queue.go
Comment (cmacknz) [inline libbeat/publisher/queue/queue.go:49]: Doesn't look like QueueLag was actually defined here.

File: libbeat/publisher/queue/queue.go
Comment (cmacknz) [reply libbeat/publisher/queue/queue.go:38]: This is a good point about what full means. It seems like just having a counter for events that were rejected because they wouldn't fit is what we'd want.

Comment (fearful-symmetry) [reply libbeat/publisher/queue/queue.go:38]: Yah, that's a good point about "full" too. Perhaps leave that to the metrics code to derive,  and let this interface just report counts and byte values, as you suggested.

File: libbeat/publisher/queue/queue.go
Comment (fearful-symmetry) [reply libbeat/publisher/queue/queue.go:37]: Yah, that's a good point, I think. When I started I was super on the fence about how much of the work of "deriving" certain values should be done by the queue, or by the metrics code.

**Comment by fearful-symmetry** [comment]
So, made some super basic changes based on what @faec said.

There's still a few missing pieces here that I want input on before we expand the metrics here. The first is how aggressive we want to get with the metrics reported through this particular interface, since there's a lot of queue-implementation-specific metrics we could report, but I'm also not sure what additional metrics we could add that would make sense with both the memory and disk queue. The memory queue isn't particularly well documented, so I'm still figuring out how it works and trying to compare it to the disk queue.

A lot of metrics we might want would also require us to actually instrument the code with additional counters, as opposed to just passing along channel sizes or user configs. I'm not sure how much of that we need/want to tackle now.


I'm also not sure if we want to implement the "queue lag" and "oldest item in queue" metrics in the queue, or if the queues should pass along some kind of internal event channel to the metrics code that we can use to track what's moving through the queue at various intervals. It's easy to imagine some kind of channel that reports things like "event with this ID and timestamp just pass through the consumer/producer" down to the metrics code, but that might be more complex than having the queue figure that out itself.

**Comment by cmacknz** [comment]
> I'm also not sure if we want to implement the "queue lag" and "oldest item in queue" metrics in the queue, or if the queues should pass along some kind of internal event channel to the metrics code that we can use to track what's moving through the queue at various intervals

One thing we will need to do is assign elements in the queue a monotonic identifier or sequence number. This will be necessary to correctly acknowledge messages between the shipper and input process (see https://github.com/elastic/elastic-agent-shipper/issues/9 for this work). Once we have that, queue lag is just the different between the sequence numbers of the newest and oldest item in the queue.

It may make sense to add those sequences numbers first if it makes the metrics easier to implement. I feel that a queue lag metric is very important as it will allow us to immediately see if the beat is being backpressued by looking at the lag metric over time. Identifying backpressure is something we really struggle with right now.

**Comment by fearful-symmetry** [comment]
> Once we have that, queue lag is just the different between the sequence numbers of the newest and oldest item in the queue.

Yah, an incrementing queue ID value would simplify a lot of this, agreed; it also means (at least to me) that deriving some of these values are simple enough that it can be done in the queue, and the metrics code shouldn't need to worry about handling some kind of event stream, or...something.

**Comment by cmacknz** [comment]
I've split off adding the queue ID / sequence numbers into a separate ticket: https://github.com/elastic/elastic-agent-shipper/issues/27

It seems like it will make sense to add that first since it simplifies the metrics implementation.

File: libbeat/publisher/queue/queue.go
Comment (cmacknz) [inline libbeat/publisher/queue/queue.go:45]: This probably requires us to implement https://github.com/elastic/beats/issues/30069 to get. Doing this for the memory queue might be tricky.

I like these and think they are a much more intuitive way to think about the queue size, but I'm not convinced adding this is more important than a lot of the other work we are already doing.

Comment (fearful-symmetry) [reply libbeat/publisher/queue/queue.go:45]: Yah, hence why I made all these `opt` values; the thinking was that some queues might care more about bytes, and others would be more event-based, the `opt` type would allow this one struct to forward whatever metrics types are available.

File: libbeat/publisher/queue/queue.go
Comment (cmacknz) [reply libbeat/publisher/queue/queue.go:39]: I think that makes sense. The queue itself doesn't care if something can fit into it, that is a concern of whoever is using the Producer interface.

File: libbeat/publisher/queue/queue.go
Comment (cmacknz) [reply libbeat/publisher/queue/queue.go:45]: OK cool, they can just stay unimplemented then until we add support for them.

Comment (faec) [reply libbeat/publisher/queue/queue.go:45]: I was thinking of these fields as referring to the queue's main buffer, which for the disk queue would mean disk usage. That data is already available in the current code, there's just no one to report it to. The memory queue could leave those fields zero for now, since it doesn't enforce a byte limit.

(We could consider adding metrics about memory overhead outside the main buffer too, but like you point out nobody tracks that yet.)

Comment (fearful-symmetry) [reply libbeat/publisher/queue/queue.go:45]: @faec Do you think this is something we should merge now? Should this or a follow-up PR actually add a `Metrics()` method to the actual interfaces?

**Comment by faec** [comment]
> It seems like it will make sense to add that first since it simplifies the metrics implementation.

I'm not sure adding that first is best, partly because I'm already working on that part and it is painfully complicated 😅 I've added an outline of the relevant state to https://github.com/elastic/elastic-agent-shipper/issues/27.

It seems like there are still a fair number of things we can immediately report or get started on, so I think we should proceed with those while the memory queue is refactored into a shape where it will handle our data. Once https://github.com/elastic/beats/issues/31307 is checked in the id change should become a lot more feasible.

**Comment by cmacknz** [comment]
Thanks for clarifying. Let's split each of the metrics into separate issues since some of them are blocked and some of them aren't:

- One issue for the event count and limit metrics. Unblocked.
- One issue for the queue lag. Requires https://github.com/elastic/elastic-agent-shipper/issues/27 which is currently blocked
- One issue for the longest waiting item metric. Unblocked.
- One issue for the byte count and limit metrics, noting that the disk queue can implement these reasonably but the memory queue will be left unimplemented initially. Unblocked for the disk queue.
- One issue for the “full queue” metric describing that it needs to be tracked by the producer (for the shipper this is the gRPC handler). Probably blocked until the gRPC handlers exist.

File: libbeat/publisher/queue/queue.go
Comment (faec) [inline libbeat/publisher/queue/queue.go:51]: What does "position" mean here? If it means byte ( / event) index in the queue buffer then this is very similar to `ByteCount` / `EventCount` (maybe modulo some rounding depending on how e.g. the disk queue reports events that are acknowledged but not yet deleted).

I'd rather leave this one off for now. I'd like things in the global interface to have an unambiguous meaning that would be consistently derivable for any queue, at least in principle, and I don't think this one is yet. If we could define it in language that is readily convertable to a unit test I'd feel better :-)

File: libbeat/publisher/queue/queue.go
Comment (faec) [inline libbeat/publisher/queue/queue.go:74]: I think this is a reasonable place to put it.

File: libbeat/publisher/queue/queue.go
Comment (fearful-symmetry) [reply libbeat/publisher/queue/queue.go:51]: So, my understanding is that this will be a measure of the "delay" involved in an event entering then exiting a queue, as told by the event IDs. Since the IDs increment, if the event leaving the queue for the output has `ID = 10` and the event entering has `ID =15`, the QueueLag is 5.

File: libbeat/publisher/queue/queue.go
Comment (fearful-symmetry) [reply libbeat/publisher/queue/queue.go:74]: Alright. Should this PR actually add this to the interface and made some dummy methods for the queues, or should we save that for when we actually implement them?

File: libbeat/publisher/queue/queue.go
Comment (faec) [reply libbeat/publisher/queue/queue.go:51]: Hmm, but if the event leaving the queue has id 10 and the event entering has id 15, then isn't that also the same as `EventCount` being 5? (+- 1 anyway) It seems redundant with that definition.

File: libbeat/publisher/queue/queue.go
Comment (faec) [reply libbeat/publisher/queue/queue.go:74]: I'd be fine either way, as long as the comments explain the intention. Dummy methods that just return an "unimplemented" error would be fine in this PR so you can start connecting the reporting components, and we can flesh out the queue-specific implementations as more pieces get connected.

File: libbeat/publisher/queue/queue.go
Comment (fearful-symmetry) [reply libbeat/publisher/queue/queue.go:51]: Yah, I'm probably describing it wrong, and discussion of this is spread out across like, 3 different issues, which isn't helping. From Craig's earlier comment:

> queue lag is just the difference between the sequence numbers of the newest and oldest item in the queue.

**Comment by fearful-symmetry** [comment]
@faec added bare methods to the queues. Hopefully we're about done.

File: libbeat/publisher/queue/queue.go
Comment (faec) [reply libbeat/publisher/queue/queue.go:52]: I don't know, to me that still sounds the same? I don't see how it isn't redundant with `EventCount`. Is there any situation where this should be different than "the number of items in the queue"?

**Review by faec — APPROVED** [review]

**Pull request merged**
